pilot/pkg/serviceregistry/util/workloadinstances/util.go

```diff
@@ -0,0 +1,87 @@
+// Copyright Istio Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package workloadinstances
+
+import (
+	"strings"
+
+	"k8s.io/apimachinery/pkg/types"
+
+	"istio.io/istio/pilot/pkg/model"
+	"istio.io/istio/pilot/pkg/util/sets"
+)
+
+// FindInstance returns the first workload instance matching given predicate.
+func FindInstance(instances []*model.WorkloadInstance, predicate func(*model.WorkloadInstance) bool) *model.WorkloadInstance {
+	for _, instance := range instances {
+		if predicate(instance) {
+			return instance
+		}
+	}
+	return nil
+}
+
+// InstanceNameForProxy returns a name of the workload instance that
+// corresponds to a given proxy, if any.
+func InstanceNameForProxy(proxy *model.Proxy) types.NamespacedName {
+	parts := strings.Split(proxy.ID, ".")
+	if len(parts) == 2 && proxy.ConfigNamespace == parts[1] {
+		return types.NamespacedName{Name: parts[0], Namespace: parts[1]}
+	}
+	return types.NamespacedName{}
+}
+
+// GetInstanceForProxy returns a workload instance that
+// corresponds to a given proxy, if any.
+func GetInstanceForProxy(index Index, proxy *model.Proxy, proxyIP string) *model.WorkloadInstance {
+	if !sets.NewSet(proxy.IPAddresses...).Contains(proxyIP) {
```

Comment:
nit: might be faster after L56 to avoid so many map creations

Reply:
replaced `sets.NewSet()` with `slices.ContainsString()`